### DISCUSSION TITLE
Transcoding HEVC

### DISCUSSION THREAD

**Discussion by marvelloard** [koush/scrypted#337] [Q&A] [answered]
Hi!

As my cameras only output HEVC, it would be a real benefit to switch to HW encoding (VAAPI). I'm running a Synology NAS (DS918+), VAAPI works. Docker container has access to the necessary devices. But what are the necessary settings (transcoder)?

**Comment by koush** [comment, marked as answer]
There are two spots you need to set up for transcoding as transcoding is a two step process, `Decoding` and `Encoding`:

## Decoding

This is configured per camera (different cameras may output different codecs).
* Enable Transcoding on the Camera's `Integrations and Extensions`. Select the decoder or enter the arguments manually.
* Different hardware decoders will have different arguments. The defaults should work, but the VAAPI arguments are for h264. Unsure if it will work for hevc.

<img width="572" alt="image" src="https://user-images.githubusercontent.com/73924/184044763-1d8ccb8d-485f-4569-996d-4097ec147481.png">


## Encoding (H264)

This is configured globally on the Rebroadcast Plugin's Transcoding settings.
* Select the h264 encoder or enter the arguments manually.
* Different hardware encoders will have different arguments.

<img width="920" alt="image" src="https://user-images.githubusercontent.com/73924/184044856-9f78923a-7b96-487e-a6d7-0af3dead04a9.png">

## FFmpeg

If the provided ffmpeg does not come with the necessary hardware encoder/decoder support, which is common for proprietary codecs, Scrypted can  use a custom ffmpeg with the environment variable `SCRYPTED_FFMPEG_PATH`.

**Reply by marvelloard** [reply]
Thank you a lot! Extremely helpful!

I found out the Synology ffmpeg version doesn't have VAAPI support compiled in. But I was able to install a different binary which is fine. Is there a way to point Scrypted to that different binary?

**Comment by koush** [comment]
scrypted comes with ffmpeg that supports vaapi. it doesn't use system ffmpeg.

**Reply by marvelloard** [reply]
The Docker container has all the necessary rights, but as soon as I set VAAPI arguments, the video doesn't load.
How can I debug this?

**Reply by koush** [reply]
the console logs for the camera will have the ffmpeg output.

**Reply by marvelloard** [reply]
Thank you! I think this is the relevant line:
`[HomeKit]: Unknown encoder 'h264_vaapi'`

Edit:
I'm afraid the supplied ffmpeg doesn't come with vaapi support.

`# ./ffmpeg -encoders
ffmpeg version 5.0.1-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2022 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)
  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gmp --enable-libgme --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libdav1d --enable-libxvid --enable-libzvbi --enable-libzimg`

**Comment by koush** [comment]
you can use the environment variable `SCRYPTED_FFMPEG_PATH` for a custom ffmpeg.

**Comment by marvelloard** [comment]
Linking ffmpeg with VAAPI statically is impossible. Can I use a custom ffmpeg outside of the container? Or could you possibly supply a docker image with ffmpeg/VAAPI in the future?

EDIT
apt install ffmpeg inside the container left me with a binary with VAAPI support!
But even with the environment variable it appears not to be used...?
Also, I guess the solution is not permanent, as the container gets cleared during an update and i loose this ffmpeg version?

**Comment by koush** [comment]
the file must exist. it works.
```ts
      const f = process.env.SCRYPTED_FFMPEG_PATH;
        if (f && fs.existsSync(f))
            return f;
```

**Comment by marvelloard** [comment]
[HomeKit]: Unknown encoder 'h264_vaapi'
[HomeKit]: ffmpeg exited

SCRYPTED_FFMPEG_PATH is set to /var/packages/VideoStation/target/bin/ffmpeg
Set binary is definitely VAAPI enabled.

**Comment by koush** [comment]
needs to be set in the container.

**Comment by marvelloard** [comment]
Almost there, thank you.
`Impossible to convert between the formats supported by the filter 'Parsed_hwupload_1' and the filter 'auto_scaler_0'`

Maybe helpful?
https://github.com/Sunoo/homebridge-camera-ffmpeg/issues/1223

**Comment by marvelloard** [comment]
I gave up.
If there would be a simple solution I would be very thankful.

**Reply by VCTGomes** [reply]
I tried to replace the current Scrypted ffmpeg installed on /root/.scrypted/node_modules/ffmpeg-static/ffmpeg to my ffmpeg with vaapi enabled and seems to have had a breakthrough. I stopped receiving the error Unknown encoder 'h264_vaapi', but now I'm receiving error Impossible to convert between the formats supported by the filter 'Parsed_scale_0' and the filter 'auto_scale_0'

Even trying to use different outputs on ffmpeg, like -vf 'format=nv12,hwupload' -c:v h264_vaapi, the erro changes and it says the format=nv12,hwupload is unknown.

**Comment by marvelloard** [comment]
I was there already as you can read here on GitHub. This is where I failed.

`Almost there, thank you.
Impossible to convert between the formats supported by the filter 'Parsed_hwupload_1' and the filter 'auto_scaler_0'

Maybe helpful?
https://github.com/Sunoo/homebridge-camera-ffmpeg/issues/1223`

**Reply by VCTGomes** [reply]
Oh, what a pity.

I really make no idea what’s going on. Same filter, using the command ffmpeg to convert a simple file, work perfectly!

**Reply by VCTGomes** [reply]
Yeah. I tried absolutely everything I could do, but no success. I tried to use Homebridge Ffmpeg and I didn't have any problems.

I'm not sure, but I suspect the problem is on command -filter_complex Plex give. 

I really make no idea what's going on.

**Reply by marvelloard** [reply]
So it works for you in homebridge via HW? Can you give me your config?

**Reply by VCTGomes** [reply]
Yeah, it does. Actually, I could setup a workaround on Scrypted to do HW transcoding right now.
 
Instead of using setting up transcoding and decoding as Koush says above, changes the Stream to Ffmpeg (TCP) and insert the commands there.

<img width="444" alt="image" src="https://github.com/koush/scrypted/assets/2366600/a5c1227f-5ead-425d-9193-870868113f73">

Commands are:
`-vaapi_device /dev/dri/renderD128
`
`-filter:v format=nv12|vaapi,hwupload -codec:v h264_vaapi -f rtsp 
`
This way Scrypted won't send -filter_complex scale, which is broking our ffmpeg. Instead of, Scrypted will use -filter:v format=nv12|vaapi,hwupload, that works!

Doing it, I'm finally able to do transcoding using GPU.
<img width="1002" alt="image" src="https://github.com/koush/scrypted/assets/2366600/b0b5d135-9700-4e12-b6ad-0074b36544a1">

However, I'm with a high CPU consumption for GPU-only transcoding. If it works for you, please tell me about the CPU usage too.

**Reply by marvelloard** [reply]
Could you please contact me privately? w.m.bauer AT gmx DOT at
I could really need some help, and this isn't the right place...

**Reply by VCTGomes** [reply]
Yeah. I can, what's your Discord?

**Reply by marvelloard** [reply]
Thank you, but surprisingly I got my problem solved after asking you!

**Reply by VCTGomes** [reply]
Oh, good to know! Are you using this workaround?

**Reply by marvelloard** [reply]
Actually no since I stay on Homebridge, as it fits my requirements better (other integrations too). This is why I wanted to contact you directly, since we're off topic.

**Comment by marvelloard** [comment]
Thank you, I will give it a try!
No, the CPU usage is fine. Maybe decoding is happening in SW, but in my experience on low power CPUs you'll always see some CPU usage in HW encoding.

**Comment by Michelinu** [comment]
Hello, can you please share how did you solve the error `Impossible to convert between the formats supported by the filter 'Parsed_scale_0' and the filter 'auto_scale_0'`

**Comment by denisgabriel5** [comment]
> There are two spots you need to set up for transcoding as transcoding is a two step process, `Decoding` and `Encoding`:
> 
> 
> 
> ## Decoding
> 
> 
> 
> This is configured per camera (different cameras may output different codecs).
> 
> * Enable Transcoding on the Camera's `Integrations and Extensions`. Select the decoder or enter the arguments manually.
> 
> * Different hardware decoders will have different arguments. The defaults should work, but the VAAPI arguments are for h264. Unsure if it will work for hevc.
> 
> 
> 
> <img width="572" alt="image" src="https://user-images.githubusercontent.com/73924/184044763-1d8ccb8d-485f-4569-996d-4097ec147481.png">
> 
> 
> 
> 
> 
> ## Encoding (H264)
> 
> 
> 
> This is configured globally on the Rebroadcast Plugin's Transcoding settings.
> 
> * Select the h264 encoder or enter the arguments manually.
> 
> * Different hardware encoders will have different arguments.
> 
> 
> 
> <img width="920" alt="image" src="https://user-images.githubusercontent.com/73924/184044856-9f78923a-7b96-487e-a6d7-0af3dead04a9.png">
> 
> 
> 
> ## FFmpeg
> 
> 
> 
> If the provided ffmpeg does not come with the necessary hardware encoder/decoder support, which is common for proprietary codecs, Scrypted can  use a custom ffmpeg with the environment variable `SCRYPTED_FFMPEG_PATH`.

is this solution still valid? seems like there is no way to set up transcoding on the latest versio. wonder why it isn't.

**Comment by crash0verride11** [comment]
What worked for me on a unifi camera sending h265 that I needed to turn into h264 for HomeKit - running Scrypted in docker on a synology NAS (ds918+):

- add `SCRYPTED_FFMPEG_PATH=/root/.scrypted/node_modules/ffmpeg-static/ffmpeg` to the docker config environment variables
- add `/dev/dri:/dev/dri` to devices or use `--device=/dev/dri:/dev/dri` to expose the gpu to the docker container
- in scrypted I set the ffmpeg input flags to `-hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format vaapi` but  `renderD128` may be different on your device
- in scrypted I set the ffmpeg output flags to `-vaapi_device /dev/dri/renderD128  -c:v h264_vaapi -bf 0 -g 48 -r 24` but the r 24 is my camera's max frame rate and the 48 is an easy multiple of that.
